x-pack/plugins/encrypted_saved_objects/server/lib/audit_logger.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export class EncryptedSavedObjectsAuditLogger {
+  constructor(private readonly enabled: boolean, private readonly auditLogger: any) {}
+
+  public encryptAttributeFailure(attributeName: string, type: string, id?: string) {
+    if (!this.enabled) {
+      return;
+    }
+
+    this.auditLogger.log(
+      'encrypt_failure',
+      `Failed to encrypt attribute "${attributeName}" for saved object "${type}:${id}".`,
+      { id, type, attributeName }
+    );
+  }
+
+  public decryptAttributeFailure(attributeName: string, type: string, id: string) {
+    if (!this.enabled) {
+      return;
+    }
+
+    this.auditLogger.log(
+      'decrypt_failure',
+      `Failed to decrypt attribute "${attributeName}" for saved object "${type}:${id}".`,
+      { id, type, attributeName }
+    );
+  }
+
+  public encryptAttributesSuccess(attributesNames: ReadonlySet<string>, type: string, id?: string) {
+    if (!this.enabled) {
+      return;
+    }
+
+    const attributesNamesArray = Array.from(attributesNames);
```

Comment:
__note:__ probably not the most efficient thing to do for every encrypt call, but hard to tell how big is the problem in real scenarios.

Reply:
Good. In situations like this I'm somewhat concerned about spending time and resources on "semi-useful-work" (not the most important part of audit logging) comparing to "useful work" (encryption) and only in case of bulk creation (that probably won't be used too much with the objects with encrypted attributes), but yeah, I agree.

Reply:
Yeah, that's a good point. Sometimes I side too strongly with avoiding premature optimization, and I definitely don't want to encourage us being wasteful.